content/courses/Ada_For_The_C_Embedded_Developer/chapters/04_Embedded.rst

```diff
@@ -240,9 +240,194 @@ Dealing with Absence of FPU with Fixed Point
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Embedded.Volatile_Object_Ada
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Volatile_Object is
+       Val : Long_Float with Volatile;
+    begin
+       Val := 0.0;
+       for I in 0 .. 999 loop
+          Val := Val + 2.0 * Long_Float (I);
+       end loop;
+
+       Put_Line ("Val: " & Long_Float'Image (Val));
+    end Show_Volatile_Object;
+
+In this example, :ada:`Val` has the :ada:`Volatile` aspect, which makes the
+object volatile. We can also use the :ada:`Volatile` aspect in type
+declarations. For example:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Embedded.Volatile_Type
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Volatile_Type is
+       type Volatile_Long_Float is new Long_Float with Volatile;
+
+       Val : Volatile_Long_Float;
+    begin
+       Val := 0.0;
+       for I in 0 .. 999 loop
+          Val := Val + 2.0 * Volatile_Long_Float (I);
+       end loop;
+
+       Put_Line ("Val: " & Volatile_Long_Float'Image (Val));
+    end Show_Volatile_Type;
+
+Here, we're declaring a new type :ada:`Volatile_Long_Float` based on the
+:ada:`Long_Float` type and using the :ada:`Volatile` aspect. Any object of this
+type is automatically volatile.
+
+In addition to that, we can declare components of an array to be volatile. In
+this case, we can use the :ada:`Volatile_Components` aspect in the array
+declaration. For example:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Embedded.Volatile_Array_Components
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Volatile_Array_Components is
+       Arr : array (1 .. 2) of Long_Float with Volatile_Components;
+    begin
+       Arr := (others => 0.0);
+
+       for I in 0 .. 999 loop
+          Arr (1) := Arr (1) +  2.0 * Long_Float (I);
+          Arr (2) := Arr (2) + 10.0 * Long_Float (I);
+       end loop;
+
+       Put_Line ("Arr (1): " & Long_Float'Image (Arr (1)));
+       Put_Line ("Arr (2): " & Long_Float'Image (Arr (2)));
+    end Show_Volatile_Array_Components;
+
+Note that it's possible to use the :ada:`Volatile` aspect for the array
+declaration as well:
+
+[Ada]
+
+.. code-block:: ada
+
+    Arr : array (1 .. 2) of Long_Float with Volatile;
+
+Atomic
+~~~~~~
+
+Operations are said to be atomic when they can be completed without
+interruptions. This is an important requirement when we're performing
+operations on objects in memory that are shared between multiple processes.
+An atomic object is an object that only accepts atomic operations. The Ada
+standard specifies that "for an atomic object (including an atomic component),
+all reads and updates of the object as a whole are indivisible."
+
+Atomicity is important, for example, when dealing with shared hardware
+registers. In Ada, we can use the :ada:`Atomic` aspect to indicate that an
+object is atomic. This is how we can use the aspect to declare a shared
```

Comment:
Yes, but not in all cases. Only when the hardware requires it, in other words. Atomicity isn't an intrinsic property of memory-mapped registers. There certainly are cases, and some are in the STM32 register set, but that's not always the case. The SVD code generator just assumes the worst and makes them all as such.

Reply:
Thanks! I adapted the paragraph accordingly.